boring/Cargo.toml

```diff
@@ -16,20 +16,38 @@ features = ["rpk", "pq-experimental"]
 rustdoc-args = ["--cfg", "docsrs"]
 
 [features]
-# Use a FIPS-validated version of boringssl.
+# Use the FIPS-validated version of boringSSL. This feature is incompatible
+# with any feature that requires (patches to) the latest version of boringSSL.
 fips = ["boring-sys/fips"]
 
-# Link with precompiled FIPS-validated `bcm.o` module.
-fips-link-precompiled = ["fips", "boring-sys/fips-link-precompiled"]
+# Link with a precompiled FIPS-validated `bcm.o` module in order to use the
+# FIPS-validated implementations of key exchange and encryption algorithms.
+fips-link-precompiled = ["boring-sys/fips-link-precompiled"]
```

Comment:
This won't work: we enabled `fips` here to enable safe API that's under `fips` flag and becomes available when we link precompiled. 

If we want to separate the two we would need to go over the safe code and replace `cfg(feature = "fips")` with `cfg(any(feature = "fips", feature = "fips-link-precompiled"))`. 

Same applies to the build script: places where we had `cfg(all(not(feature = "fips"), feature = "fips-link-precompiled")` need to be replaced with just `cfg(feature = "fips-link-precompiled")` and places where we have just `cfg(feature = "fips")` need to be changed to `cfg(any(feature = "fips", feature = "fips-link-precompiled"))`

Reply:
Hmm, so the reason I made this change is that I noted the following when I try to build with "fips-link-precompiled" (this is reproducible on `master`):


```
chris@ubutron:~/github.com/cjpatton/boring$ export RUSTFLAGS="-Clink-args=-Wl,-zmuldefs" && export BORING_SSL_PRECOMPILED_BCM_O=/opt/boringssl-fips/lib/bcm.o && cargo clean && cargo test --features fips-link-precompiled
                ... snipped ...
   Compiling tokio v1.30.0
   Compiling boring-sys v3.0.4 (/home/chris/github.com/cjpatton/boring/boring-sys)
   Compiling rusty-hook v0.11.2
   Compiling futures-executor v0.3.28
   Compiling futures v0.3.28
   Compiling tokio-util v0.7.8
   Compiling h2 v0.3.20
   Compiling hyper v0.14.27
warning: linking in precompiled bcm.c.o` module
   Compiling boring v3.0.4 (/home/chris/github.com/cjpatton/boring/boring)
error: could not find native static library `crypto`, perhaps an -L flag is missing?

The following warnings were emitted during compilation:

warning: linking in precompiled bcm.c.o` module

error: could not compile `boring-sys` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
error[E0308]: mismatched types
    --> boring/src/bio.rs:33:17
     |
31   |             cvt_p(BIO_new_mem_buf(
     |                   --------------- arguments to this function are incorrect
32   |                 buf.as_ptr() as *const _,
33   |                 buf.len() as BufLen,
     |                 ^^^^^^^^^^^^^^^^^^^ expected `isize`, found `i32`
     |
note: function defined here
    --> /home/chris/github.com/cjpatton/boring/target/debug/build/boring-sys-f0b0def01807e627/out/bindings.rs:9461:12
     |
9461 |     pub fn BIO_new_mem_buf(buf: *const ::std::os::raw::c_void, len: ossl_ssize_t) -> *mut BIO;
     |            ^^^^^^^^^^^^^^^
help: you can convert an `i32` to an `isize` and panic if the converted value doesn't fit
     |
33   |                 (buf.len() as BufLen).try_into().unwrap(),
     |                 +                   +++++++++++++++++++++

error[E0308]: mismatched types
     --> boring/src/ssl/mod.rs:1316:17
      |
1313  |             let r = ffi::SSL_CTX_set_alpn_protos(
      |                     ---------------------------- arguments to this function are incorrect
...
1316  |                 protocols.len() as ProtosLen,
      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `usize`, found `u32`
      |
                       ... snipped ...
```

Reply:
This happens because the rust bindings are no longer compatible with `ffi` with the version of boring in boring-sys/dep/boringssl-fips.

Reply:
See #147.